### ISSUE TITLE
[FEATURE] Apply schema default values immediately on instantiation of a document

### ISSUE THREAD

**Issue by ablankenship10** [dynamoose/dynamoose#905]
### Proposal:

- Add dynamoose.model.defaults key of `createWithDefaults` (default false to make a non breaking change)

- Set to true, your defaults would be applied immediately on creating a new document whether its saved or not.

- In the case a user modifies the document breaking the defaults which have already been applied, thatd be a user error and the call to save would either fail or succeed with their intentionally changed document.

#### Schema
```javascript
dynamoose.model.defaults = { createWithDefaults: true };

const MyModel= dynamoose.model('MyModel', {
	id: {
		type: String,
		default: () => uuid(),
		hashKey: true,
	},
	email: {
		type: String,
		required: true,
	},
	verified: {
		type: Boolean,
		required: true,
		default: false,
	},
});

const modelInstance = new MyModel({ email: 'test@example.com'});
const modelInstance = await MyModel.create({ email: 'test@example.com'});

/* 

Should both produce:

{ id: '9981f6d0-92ca-41b8-9c44-2c1d09effd3b', email: 'test@example.com', verified: false }

Regardless of the document being saved/created to the DB or not.

Instead of just:

{ email: 'test@example.com' }

*/

```

**Comment by ablankenship10** [comment]
@fishcharlie 
Spent some time digging into how this could work. It seems like the main difficulty is being able to run `Document.conformToSchema()` within the constructor of the document. Being async obviously it cant run there. For the most part I attempted to remove all the async calls that seemed unnecessary but I assume they're there in the case a user has default value functions that are async? Either way I broke about every test after this so I'm thinking this is going to require a bit more thought.

I wouldn't be opposed to running something like `conformToSchema()` after running `new MyModel({...})`, however the function as of now has a lot of parameters required that isn't very user friendly to use inline. Would need some type of helper wrapper around it to produce the desired result.

**Comment by ablankenship10** [comment]
Tried another approach that can work for optionally returning the new document values after a save. Worked out well and didn't break a single test :) Let me know what you think!

**Comment by ablankenship10** [comment]
One bug in this approach found so far is that Documents with timestamps return the time in epoch ms, rather than date objects.

**Comment by Pkthunder** [comment]
Have these changes been merged or require further testing?

**Comment by daria-garkusha** [comment]
Is there any progress with this feature? It's quite critical for us because we need to return back a document with default values after save or create. Currently, as a workaround we execute create method with { 'return': 'request' } settings to prefill default values and then save the document with all attributes. But it would be nice to get the correct document back by default or with an extra setting.

**Comment by IAmKio** [comment]
Keen on this too - for the same reasons as @daria-garkusha.

**Comment by fishcharlie** [comment]
I'm really not sure what the best way to handle this is.

1. Have it apply default values by default no matter what (would be a breaking change).
    - Downside to this would be it needs a breaking change
2. Have a setting on the Model to use default values when creating documents. (Sounds like this is what the original proposal was for)
    - Downside to this, is what if you only want some documents for a given model to use default values by default? It's not very flexible or customizable.
3. Have a setting on the Document to use default values when creating documents.
    - Downside to this is that it's not model wide. So kinda the reverse of number 2, what if you want it to have it happen for all models? One way to solve this is to allow the setting to be on both Document and Model.
4. Have a method on Document to apply default values. (ex. `document.applyDefaults()`)
    - Personally I'm a fan of this option, just because it's explicit and doesn't have side effects in already existing code. However, calling this a bunch of times gets really strange. So having these spread out throughout your project will get messy quickly.
5. Maybe more that I'm not thinking of.

Also, what happens if you create a document/item, and then remove a default value, then try to save it. How should Dynamoose react to that?

Maybe the solution here is a combination of the ideas listed. Just not sure. Any feedback or thoughts here would be appreciated.

**Comment by Pkthunder** [comment]
One thing to consider is currently the default attribute allows `async` functions, which complicates defaults being set on constructed Documents

```javascript
dynamoose.model.defaults = { createWithDefaults: true };

const MyModel = dynamoose.model('MyModel', {
	id: {
		type: String,
		default: () => uuid(),
		hashKey: true,
	},
	age: {
                 type: Number,
                 default: async () => {
                         const networkResponse = await axios("https://myurl.com/config.json").data;
                         return networkResponse.defaults.age;
                  }
        }
});

const modelInstance = new MyModel({ email: 'test@example.com'});
// { id: "1235.....", age: Promise<Pending> }

const modelInstance = await MyModel.create({ email: 'test@example.com'});
// { id: "1235.....", age: 25 }
```

Perhaps the approach is to add a setting to `.create`, so a document can be created like so: 

```javascript
MyModel.create({id: 1, name: 'foo'}, { withDefaults: true });
```
This seems a lower-stakes change and avoids the complexities of breaking changes and async operations within a constructor. This could also be something that can be configured/overridden at a Model or Document level.

Thoughts?

**Comment by fishcharlie** [comment]
@Pkthunder Totally forgot about that. Great point.

My issue with adding another setting, is to me in this case, that doesn't seem like the best long term solution. The reality is `document.save` & `Model.create` should return a promise that gets resolved with the document that was saved (including default values). Currently in the documentation it doesn't mention what `document.save` or `Model.create` will return.

> This method returns a promise that will resolve when the operation is complete, this promise will reject upon failure.

So an argument could be made that changing that behavior isn't truly a breaking change.

However, that is **really** different than `new Model()`. So that solution although it works for save and create, won't work for just creating a new instance. This is where I get back to having a function on Document that can set the default values, might be an interesting idea.

So to me, this is almost a two factor thing.

1. Result from save or create should return saved item/document. To me, this is clear, explicit, and should be done. Obvious clear path forward here. The only question is do we consider this a breaking change and how does that fit into the release cycle?
2. How do we handle the case of creating an item/document without actually saving it? This is where more questions are raised. I'm more of a fan of just having a method on the document that you can call to set the defaults. That way async stuff can work fine.

To me, when you really think about it `MyModel.create({id: 1, name: 'foo'}, { withDefaults: true });` doesn't make sense. Is there **ever** a time where you don't want to have `withDefaults` set to false? Arguably no. This shouldn't be an option. The behavior should just be to return the actual created document with defaults.

Now. Let's pretend you have a weird case where you do want it to be separate. What's preventing you from doing the following then?

```js
const document = new MyModel({id: 1, name: 'foo'}); // document without default values
const newDocument = await document.save({overwrite: false}); // document with default values
```

So I do think that flexibility is inherently built in to where having a `withDefaults` option is unnecessary. It's just the behavior needs to be changed IMO.

The only value in adding `withDefaults` setting, to me, is it's a super quick hack. It gets the job done and we could ship a release quickly. However, I don't think the quick solution in this case is the best solution. I think we as a project should strive for the best solution, and especially in this case I think having a breaking change or something is the right path to go down, as opposed to a quick workaround.

**Comment by bladimir-villarreal** [comment]
Hi, i'm here because i need the Id (hashKey) of my schema.
there is
`const dynamoose = require('dynamoose');
const { v4: uuidv4 } = require('uuid');
var Schema = dynamoose.Schema;

var dataOriginSchema = new Schema({
    Id: {
        type: String,
        hashKey: true,
        default: uuidv4()
    },
    Name: String,
    TypeOrigin: String,
    BucketName: String,
    Separator: String,
    CreatedDate: String,
}, {
    saveUnknown: true,
});`

And when i save my document i don't have the Id in the response
`
async function create(RequestDataO) {
    const validDataOrigin = await DataOrigin.scan('Name').contains(RequestDataO.Name).exec();
    if (validDataOrigin[0]) {
        throw 'Name ' + RequestDataO.Name + ' Already Exists';
    }

    RequestDataO.CreatedDate = new Date().toString();
    const dataOrigin = new DataOrigin(RequestDataO);
    return await dataOrigin.save();
}`

**Comment by jdrydn** [comment]
@bladimir-villarreal So far the only way I've been achieving that is by creating an object by hand each time before passing to Dynamoose, so either:

```js
const dog = {
  id: uuid(),
};

const dogModel = new Animal(dog);
dogModel.save();
// or
await Animal.create(dog);
```

It would be good to see Dynamoose return the completed object that it inserted into DynamoDB.

**Comment by Pkthunder** [comment]
To solve this I created a model method that will apply all defaults in the schema before `.save` is called and I use it instead of `.create`

```javascript
Animal.methods.set('createWithDefaults', function (obj) {
  const keys = Object.keys(this.schema.schemaObject).filter(key => {
    const schema = this.schema.schemaObject[key];
    return schema.default && schema._defaultFix;
  });
  return Promise.all(keys.map(key => {
    const schema = this.schema.schemaObject[key];
    return Promise.resolve()
    .then(() => schema.default())
    .then(defaultVal => {
      obj[key] = defaultVal
    });
  }))
  .then(() => {
    // console.log(obj);
    return this.create(obj);
  })
});
```

```javascript
const props = {name: 'Toto'};
const dog = await Animal.createWithDefaults(props);
```

**Comment by softmarshmallow** [comment]
Need this. expected behaviour for non-set fields with default value set, setting the default is expected. actually i think thats the only case. of default's usage. (if without force default)

**Comment by softmarshmallow** [comment]
Question here.

![image](https://user-images.githubusercontent.com/16307013/101563930-8416ba80-3a0d-11eb-8655-dde837cc51f2.png)


Is it not implemented or not working as expected?

I have very simple schema,
``` ts

const KeySchema = new dynamoose.Schema({
    id: {
        type: String,
        hashKey: true,
        default: () => nanoid()
    },
    keyName: {
        type: String,
        required: true,
    },
    embeddable: {
        type: Boolean,
        default: false,
        required: true
    }
}, {
    saveUnknown: false
})

const KEY_TABLE_NAME = process.env.DYNAMODB_KEY_TABLE
export const KeyModel = dynamoose.model(KEY_TABLE_NAME, KeySchema, {
    create: false
})
```

Which gives me no id auto generated (as well as embeddable)

**Comment by Pkthunder** [comment]
> default value for an attribute that will be applied upon _save_

Are you calling `.save` or `KeyModel.create(...)`? Your snippet does not show your use of `KeyModel`

**Comment by softmarshmallow** [comment]
Sorry, I'm using save(). should i be supposed to use create?

**Comment by Pkthunder** [comment]
Apologizes, my response was misleading. There are a few work-arounds suggested in this thread, but currently default model attribute values to not behave they way you're expecting

**Comment by softmarshmallow** [comment]
Thanks for the reply.
Alright, jus curious. what is `defualt` then?
What does it do?

**Comment by jdrydn** [comment]
Default works as you expect - it sets default properties on the instance when created/updated. The trouble is it doesn't return these values, so (for example) if you use a `uuid` function to generate IDs you don't get back that ID 🤦‍♂️ 

@Pkthunder dropped a useful model method [earlier in this thread](https://github.com/dynamoose/dynamoose/issues/905#issuecomment-708688323) which is an excellent workaround until such a time as this package fixes this 💪

**Comment by softmarshmallow** [comment]
Not returning the value.-> is this by design?

**Comment by jdrydn** [comment]
FYI as of `2.5.0` it would appear `schemaObject` is in a different location, so my implementation has shifted slightly to:

```
const fillInDefaults = async create => {
  const { schemas: { '0': { schemaObject } } } = Animal;
  await Promise.all(Object.keys(schemaObject).filter(key => {
    const { [key]: field } = schemaObject;
    return field.hasOwnProperty('default') && !create.hasOwnProperty(key);
  }).map(async key => {
    const { [key]: { default: $default } } = schemaObject;
    create[key] = typeof $default === 'function' ? (await $default()) : $default;
  }));

  return create;
};

Animal.methods.set('createWithDefaults', async create => model.create(await fillInDefaults(create)));
Animal.transaction.createWithDefaults = async create => model.transaction.create(await fillInDefaults(create));
```

I also added an `async` method to `transaction` to support transactions with defaults! 🎄

**Comment by jdrydn** [comment]
@fishcharlie This might actually be a bug? For different reasons, I ended up on the docs for [DynamoDB's putItem](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB.html#putItem-property), it can return the `Attributes` that were sent to DynamoDB:

<img width="460" alt="Screenshot 2020-12-24 at 16 58 05" src="https://user-images.githubusercontent.com/406967/103100176-5e8ed100-4609-11eb-8477-0ff6738a46fa.png">

Which suggests that the document that's returned from `model.create` or `document.save` should contain all the default values in it?

**Comment by fishcharlie** [comment]
@jdrydn I'm confused about what point you are trying to make. Read https://github.com/dynamoose/dynamoose/issues/905#issuecomment-651558108.

If after reading that you are in disagreement, or have additional comments, please leave them. I'm **really** looking for feedback here (hence why this issue is tagged with discussion).

---

Remember, this issue has gotten sidetracked in a major way. The original post says:

```
const modelInstance = new MyModel({ email: 'test@example.com'});
const modelInstance = await MyModel.create({ email: 'test@example.com'});

/* 
Should both produce:

{ id: '9981f6d0-92ca-41b8-9c44-2c1d09effd3b', email: 'test@example.com', verified: false }
```

A lot of this post is now referring to an issue where `model.create` and `document.save` don't return items with default values.

To me, these are two different issues.

1. `new Model` returning default values, I'm unconvinced if that is a good idea. I also think it raises complexity around async stuff.
2. I'm in agreement that `Model.create` and `document.save` should return the object they sent to the DB.

I almost consider item 2 to be blocked by #1059. I know they are pretty unrelated, but it touches the same logic/function, and there is a chance that the two are connected somehow.

I also consider item 2 to be unrelated to the original issue/post, tho a lot of recent conversation has focused on that item. So maybe your comment about this being a bug, is related to the tag on this issue saying `type:enhancement` or the `FEATURE` in the title? If so, that is because I consider the original issue to be related to point 1, not point 2.

**Comment by fishcharlie** [comment]
Two points I made in my comment above:

> `new Model` returning default values, I'm unconvinced if that is a good idea. I also think it raises complexity around async stuff

No further discussion has occurred around this. So I assume there is agreement with this point since there are 3 👍's to my comment with no comments afterwards.

> I'm in agreement that `Model.create` and `document.save` should return the object they sent to the DB

This was fixed in v2.7.2 #1141.

---

Therefore, I'm closing this issue, with one condition.

As part of v3 #1040, I have opened an issue to create a `item.applyDefaults` method. Not quite the same as applying defaults upon creation (`new Model`), but can achieve the same goal with one more method call. You can find this issue at: #1155.

---

If you disagree or have more to add to this issue, **please** follow the guidelines before commenting:

1. Ensure that your comment is on topic (this issue has gotten side tracked before, and would like to keep it on focus)
2. If you believe it's not on topic, please submit a new issue
3. Read #1155 and if you believe your comment is more suited to that issue, please comment there instead
4. If you believe it's on topic and directly related to the original post OR what I said in this comment, feel free to comment here

If this post gets too off topic again I will be locking this thread.

---

Hopefully this addresses everything, and I look forward to implementing #1155 for v3.
